platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -506,28 +509,76 @@ private void shrinkIfNeeded(final long newMinValidIndex) {
         }
 
         // clean up a chunk with data
-        final int firstChunkWithData = (int) (newMinValidIndex / numLongsPerChunk);
+        final int firstChunkWithDataIndex = (int) (newMinValidIndex / numLongsPerChunk);
         final long numberOfElementsToCleanUp = (newMinValidIndex % numLongsPerChunk);
-        partialChunkCleanup(firstChunkWithData, numberOfElementsToCleanUp);
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfElementsToCleanUp > 0) {
+            partialChunkCleanup(chunk, true, numberOfElementsToCleanUp);
+        }
 
         // clean up chunk(s) reserved for buffer
-        for (int i = firstValidChunkWithBuffer; i < firstChunkWithData; i++) {
-            partialChunkCleanup(i, numLongsPerChunk);
+        for (int i = firstValidChunkWithBuffer; i < firstChunkWithDataIndex; i++) {
+            chunk = chunkList.get(i);
+            if (chunk != null) {
+                partialChunkCleanup(chunk, true, numLongsPerChunk);
+            }
+        }
+    }
+
+    /**
+     * Deletes values from {@code maxValidIndex} to the end of the list and releases memory chunks reserved for these values.
+     * @param maxValidIndex new maximal valid index, right boundary of the list
+     */
+    private void shrinkRightSideIfNeeded(long maxValidIndex) {
+        final int listLength = chunkList.length();
+        final int lastValidChunkWithBufferIndex =
+                (int) Math.min((maxValidIndex + reservedBufferLength) / numLongsPerChunk, listLength);
+        final int firstChunkIndexToDelete = lastValidChunkWithBufferIndex + 1;
+        final int numberOfChunks = calculateNumberOfChunks(size());
+
+        for (int i = firstChunkIndexToDelete; i < numberOfChunks; i++) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                releaseChunk(chunk);
+            }
+        }
+
+        // clean up a chunk with data
+        final int firstChunkWithDataIndex = (int) (maxValidIndex / numLongsPerChunk);
+        final long numberOfElementsToCleanUp = numLongsPerChunk - (maxValidIndex % numLongsPerChunk) - 1;
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfElementsToCleanUp > 0) {
+            partialChunkCleanup(chunk, false, numberOfElementsToCleanUp);
+        }
+
+        // clean up chunk(s) reserved for buffer
+        for (int i = firstChunkWithDataIndex + 1; i <= lastValidChunkWithBufferIndex; i++) {
+            if (i >= listLength) {
```

Comment:
Is this really needed? `lastValidChunkWithBufferIndex` is guaranteed to not exceed `listLength` because of `Math.min()` above.

Reply:
Yes, fair enough. It should rather be `length - 1` because it's an index.